aiida_vasp/workflows/relax.py

```diff
@@ -0,0 +1,298 @@
+"""Structure relaxation workchain for VASP."""
+import numpy
+from aiida.common.extendeddicts import AttributeDict
+from aiida.work import WorkChain
+from aiida.work.workchain import append_, while_, ToContext
+
+from aiida_vasp.utils.aiida_utils import get_data_class, get_data_node
+from aiida_vasp.utils.vasp.isif import IsifStressFlags, Isif
+from aiida_vasp.utils.vasp.ibrion import IbrionFlags, Ibrion
+from aiida_vasp.utils.vasp.encut import EncutFlags, Encut
+from aiida_vasp.utils.vasp.nsw import Nsw
+from aiida_vasp.workflows.base import VaspBaseWf
+from aiida_vasp.workflows.restart import prepare_process_inputs, UnexpectedCalculationFailure
+
+RELAXATION_INCAR_TEMPLATE = AttributeDict({
+    "ismear": -1,
+    "sigma": 0.2,
+    "algo": 'NORMAL',
+    "ediff": 1E-6,
+    "prec": 'Normal',
+    "ibrion": 2,
+    "nsw": 200,
+    "nelmin": 4,
+    "ediffg": -0.01,
+    "isif": 4
+})
+
+
+def relax_parameters(pos, shape, volume):
+    """Set IBRION and ISIF according to the choice of what should be relaxed."""
+    if not (pos or volume or shape):
+        raise ValueError('relaxation workflow requires at least one of "positions", "cell shape", "cell volume" to be alowed to change.')
+    if not pos:
+        ibrion = Ibrion(ion_updates=IbrionFlags.NO_UPDATE)
+    else:
+        ibrion = Ibrion(ion_updates=IbrionFlags.IONIC_RELAXATION_CG)
+    isif = Isif(calculate_stress=IsifStressFlags.FULL, vary_positions=pos, vary_cell_shape=shape, vary_cell_volume=volume)
+    nsw = Nsw(value=200, ibrion=ibrion)
+    return ibrion, isif, nsw
+
+
+def encut_factor(volume):
```

Comment:
Let us not do this. This has to be tested. E.g. one defines an ENCUT, relaxes and checks if that ENCUT was fine. One can cheat and check other parameters so the relaxation is not necessary, e.g. the change in some parameters upon a shift of one or several atoms. etc. This means we just eat whatever ENCUT is passed for this workchain.

Reply:
So this would be done in a separate ENCUT determining workchain? or a check after each relaxation step? Could you tell me what you would check for and how you would choose the next ENCUT value to try?

Reply:
Yes, we typically check for convergence by iterating over different ENCUT and k-point meshes. Usually we start with a sensible k-point mesh (as we know more about this before than ENCUT) and do the ENCUT check. Then when we have a converged ENCUT, we fix that and iterate over the k-point densities. When both are withing some acceptable criteria we move along. There are many different ways that we do this and it usually depends on the calculations that follow and what kind of properties we want. Many people just checked against the total energy. E.g. that the total energy should be within, say a few meV. However, in recent years people have realized that this might be too strict for some calculations and one has started to do relative convergence tests. This means, if one is displacing atoms and trying to find the most stable structure one should check the convergence of the total energy difference. So basically one would have to do at least two calculations pr. convergence step for one example atom displacement (can be necessary to do more). Some people check band gaps, some dielectric function etc. I would say we start with: total energies, relative total energies based on some atomic displacement and the band gap. We (or the users) can always add more later.
Also, when we do these tests we get curves and sometimes it makes sense to parameterize and find the exact value given a criteria. For high throughput all this is super critical, because your number one concern is computational resources.